File: hackerrank/Daeseong/big_sorting_U.java
import java.io.*;
import java.math.*;
import java.security.*;
import java.text.*;
import java.util.*;
import java.util.concurrent.*;
import java.util.regex.*;

public class Solution {

    // Complete the bigSorting function below.
    static String[] bigSorting(String[] unsorted) {
        PriorityQueue<BigNum> pq = new PriorityQueue<>();
        for(int i = 0; i < unsorted.length; i++){
            pq.offer(new BigNum(unsorted[i].length(), unsorted[i]));
        }
        int index = 0;
        while (!pq.isEmpty()){
            BigNum temp = pq.poll();
            unsorted[index] = temp.str;
            index++;
        }

        return unsorted;
    }

    static class BigNum implements Comparable<BigNum>{
        int len;
        String str;

        public BigNum(int len, String str) {
            this.len = len;
            this.str = str;
        }

        @Override
        public int compareTo(BigNum o) {
            if (this.len > o.len) return 1;
            else if(this.len < o.len) return -1;
            else{
                for(int i = 0; i < o.len; i++){
                    if(this.str.charAt(i) > o.str.charAt(i)) return 1;
                    else if(this.str.charAt(i) < o.str.charAt(i)) return -1;
                }
                return 0;
            }
        }
    }

    private static final Scanner scanner = new Scanner(System.in);

    public static void main(String[] args) throws IOException {
        BufferedWriter bufferedWriter = new BufferedWriter(new FileWriter(System.getenv("OUTPUT_PATH")));

        int n = scanner.nextInt();
        scanner.skip("(\r\n|[\n\r\u2028\u2029\u0085])?");

        String[] unsorted = new String[n];

        for (int i = 0; i < n; i++) {
            String unsortedItem = scanner.nextLine();
            unsorted[i] = unsortedItem;
        }

        String[] result = bigSorting(unsorted);

        for (int i = 0; i < result.length; i++) {
            bufferedWriter.write(result[i]);

            if (i != result.length - 1) {
                bufferedWriter.write("\n");
            }
        }

        bufferedWriter.newLine();

        bufferedWriter.close();

        scanner.close();
    }
}

File: hackerrank/Daeseong/big_sorting_U.py
# Solution
# - 파이썬은 자료형의 범위가 없다. 따라서 정수형 비교가 바로 가능하다
# - 하지만 변환하지 않고 비교하는 방법을 사용할 수 있다.
# - key에 lambda를 넣어 다중 조건실을 활용했다.
# - 우선 순위는 다음과 같다.
# - 1. 길이의 오름차순
# - 2. 문자열의 오름차순

#!/bin/python3

import math
import os
import random
import re
import sys

# Complete the bigSorting function below.
def bigSorting(unsorted):
    return sorted(unsorted, key=lambda x: (len(x), x))

if __name__ == '__main__':
    fptr = open(os.environ['OUTPUT_PATH'], 'w')

    n = int(input())

    unsorted = []

    for _ in range(n):
        unsorted_item = input()
        unsorted.append(unsorted_item)

    result = bigSorting(unsorted)

    fptr.write('\n'.join(result))
    fptr.write('\n')

    fptr.close()
